sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java

```diff
@@ -82,29 +81,23 @@ public void close() { }
      * forbidden mutations.
      */
     private final T possiblyModifiedObject;
-
-    /**
-     * A saved encoded copy of the same value as {@link #possiblyModifiedObject}. Naturally, it
-     * will not change if {@link #possiblyModifiedObject} is mutated.
-     */
-    private final byte[] encodedOriginalObject;
-
     /**
-     * The object decoded from {@link #encodedOriginalObject}. It will be used during every call to
+     * The structural value from {@link #possiblyModifiedObject}.
+     * It will be used during every call to
      * {@link #verifyUnmodified}, which could be called many times throughout the lifetime of this
      * {@link CodedValueMutationDetector}.
      */
-    private final T clonedOriginalObject;
+    private final Object originalStructuralValue;
 
     /**
      * Create a mutation detector for the provided {@code value}, using the provided {@link Coder}
      * for cloning and checking serialized forms for equality.
      */
     public CodedValueMutationDetector(T value, Coder<T> coder) throws CoderException {
       this.coder = coder;
+      this.originalStructuralValue = coder.structuralValue(value);
```

Comment:
You'll want to use the structural value of the cloned object.

If something is consistent with equals, then ```structuralValue(value) == value```. Hence modifying value will go unnoticed.